src/main/scala/leon/synthesis/rules/IntegerEquality.scala

```diff
@@ -0,0 +1,421 @@
+    case APAInputCombination(coef, input_linear) =>
+      ((InfiniteIntegerLiteral(BigInt(coef)): Expr) /: input_linear) { (c: Expr, iv) =>
+        val (i, v) = iv
+        Plus(c, Times(InfiniteIntegerLiteral(i), Variable(vc.getIdentifier(v))))
+      }
+    case APAInputDivision(numerator, denominator) =>
+      Division(APAInputTermToExpr(APAInputMultiplication(numerator)), APAInputTermToExpr(APAInputMultiplication(denominator)))
+    case APAInputMultiplication(operands) =>
+      val v = (operands map APAInputTermToExpr)
+      v match {
+        case Nil => InfiniteIntegerLiteral(1)
+        case a::Nil => a
+        case l => l.reduce(Times)
+      }
+    case APAInputAddition(operands) =>
+      val v = (operands map APAInputTermToExpr)
+      v match {
+        case Nil => InfiniteIntegerLiteral(0)
+        case a::Nil => a
+        case l => l.reduce(Plus)
+      }
+    case APAInputAbs(a) =>
+      val absFun = vc.sctx.program.library.lookupUnique[FunDef]("leon.math.abs")
+      functionInvocation(absFun, Seq(APAInputTermToExpr(a)))
+    case APAInputGCD(numbers) =>
+      import vc.sctx.program.library._
+      val gcdFun = lookupUnique[FunDef]("leon.math.gcdlist")
+      val args = numbers map APAInputTermToExpr
+      val l = (args :\ (CaseClass(CaseClassType(Nil.get, Seq(IntegerType)), Seq()): Expr)) { (v, l) =>
+        CaseClass(CaseClassType(Cons.get, Seq(IntegerType)), Seq(v, l))
+      }
+      functionInvocation(gcdFun, Seq(l))
+    case APAInputLCM(numbers) =>
+      import vc.sctx.program.library._
+      val lcmFun = lookupUnique[FunDef]("leon.math.lcmlist")
+      val args = numbers map APAInputTermToExpr
+      val l = (args :\ (CaseClass(CaseClassType(Nil.get, Seq(IntegerType)), Seq()): Expr)) { (v, l) =>
+        CaseClass(CaseClassType(Cons.get, Seq(IntegerType)), Seq(v, l))
+      }
+      functionInvocation(lcmFun, Seq(l))
+  }
+}
+
+case object IntegerEquality extends Rule("Solve integer equality"){
+  
+  import ComfusyConverters._
+  
+  
+  def instantiateOn(implicit hctx: SearchContext, p: Problem): Traversable[RuleInstantiation] = {
+    implicit lazy val vc = new VarContext(p.as.toSet, p.xs.toSet, hctx)
+    p.phi match {
+      case TopLevelAnds(conjuncts) =>
+        val candidates: List[Equals] = conjuncts.toList collect { case e: Equals => e } filter { e => isLinEqInOutputVariables(e) }
+        if(candidates.isEmpty) {
+          Nil
+        } else {
+          val realCandidates  = candidates map (e => convertToAPAEqualZero(e)) sortWith APASynthesis.by_least_outputvar_coef
+          val problem = new APASynthesis(FormulaSplit(realCandidates, Nil, Stream.empty), p.as.toList.map(vc.getInputVar), p.xs.toList.map(vc.getOutputVar))
+          List(RuleInstantiation("Solve integer equalities") {
+            val (cond, prog) = problem.solve()
+            val solution = convertToSolution(cond, prog)
+            RuleClosed(solution)
```

Comment:
@manoskouk @samarion How can I say that I have a new problem instead of outputting a solution? I'm only solving equalities for now.